src/main.rs

```diff
@@ -234,37 +242,51 @@ async fn handle_request(
         }
     };
 
-    let bytes = match hyper::body::to_bytes(body).await {
-        Ok(body) => body.to_vec(),
+    let p = path_name(&path);
+
+    let header_sender = match ratelimiter.ticket(path).await {
+        Ok(sender) => sender,
         Err(e) => {
-            error!("Failed to receive incoming request body: {:?}", e);
-            return Err(RequestError::ChunkingRequest { source: e });
+            error!("Failed to receive ticket for ratelimiting: {:?}", e);
+            return Err(RequestError::AcquiringTicket { source: e });
         }
     };
 
-    let path_and_query = if let Some(v) = uri.path_and_query() {
-        v.as_str().replace(&api_url, "")
-    } else {
-        error!("No path in URI: {:?}", uri);
-        return Err(RequestError::NoPath { uri });
-    };
-    let p = path_name(&path);
-    let raw_request = RequestBuilder::raw(method, path, path_and_query)
-        .body(bytes)
-        .headers(headers.into_iter().filter_map(|(k, v)| k.map(|h| (h, v))))
-        .build();
+    parts
+        .headers
+        .insert(AUTHORIZATION, token.try_into().unwrap());
+    parts
+        .headers
+        .insert(HOST, HeaderValue::from_static("discord.com"));
+
+    api_url.push_str(&trimmed_path);
+
+    let mut uri_parts = parts.uri.into_parts();
+    uri_parts.scheme = Some(Scheme::HTTPS);
+    uri_parts.authority = Some(Authority::from_static("discord.com"));
+    uri_parts.path_and_query = PathAndQuery::from_maybe_shared(api_url).ok();
+    let new_uri = Uri::from_parts(uri_parts).unwrap();
+    parts.uri = new_uri;
+
+    let new_request = Request::from_parts(parts, body);
 
     #[cfg(feature = "expose-metrics")]
     let start = Instant::now();
 
-    let resp = match client.request::<Vec<u8>>(raw_request).await {
-        Ok(resp) => resp,
+    let resp = match client.request(new_request).await {
+        Ok(response) => response,
         Err(e) => {
-            error!("Failed to receive reply body: {:?}", e);
+            error!("Error when requesting the Discord API: {:?}", e);
             return Err(RequestError::RequestIssue { source: e });
         }
     };
 
+    let ratelimit_headers = RatelimitHeaders::try_from(resp.headers()).ok();
+    if header_sender.send(ratelimit_headers).is_err() {
+        error!("Error when sending ratelimit headers to ratelimiter");
+        return Err(RequestError::SendingRatelimitHeaders);
+    };
+
```

Comment:
we've already done the request here (and gotten some response back), shouldn't we return the response as normal? (just print the error)